pkg/kube/watcher.go

```diff
@@ -124,41 +132,89 @@ func WatchForServices(ctx context.Context, tracker *tcplistener.ListenerTracker,
 					"error": err,
 				})
 				watchCancel()
-				watchContext, watchCancel = context.WithCancel(ctx)
+
 				state = stateNoConfig
 
 				time.Sleep(time.Second)
 
 				continue
 			case event := <-eventCh:
 				if event.deleted {
-					if err := tracker.Remove(ctx, k8sServiceListenerIP, int(event.port)); err != nil {
-						log.Errorw("failed to close listener", log.Fields{
-							"error":     err,
-							"port":      event.port,
-							"namespace": event.namespace,
-							"name":      event.name,
-						})
+					if enablePrivilegedService {
+						if err := portTracker.Remove(string(event.UID)); err != nil {
+							log.Errorw("failed to delete a port from tracker", log.Fields{
+								"error":     err,
+								"UID":       event.UID,
+								"ports":     event.portMapping,
+								"namespace": event.namespace,
+								"name":      event.name,
+							})
+
+							continue
+						}
+
+						log.Debugf("kuberentes service: port mapping deleted %s/%s:%v",
+							event.namespace, event.name, event.portMapping)
 
 						continue
 					}
 
-					log.Debugf("kuberentes service: deleted listener %s/%s:%d",
-						event.namespace, event.name, event.port)
+					for port := range event.portMapping {
+						if err := tracker.Remove(ctx, k8sServiceListenerIP, int(port)); err != nil {
+							log.Errorw("failed to close listener", log.Fields{
+								"error":     err,
+								"ports":     event.portMapping,
+								"namespace": event.namespace,
+								"name":      event.name,
+							})
+
+							continue
+						}
+					}
+
+					log.Debugf("kuberentes service: deleted listener %s/%s:%v",
+						event.namespace, event.name, event.portMapping)
 				} else {
-					if err := tracker.Add(ctx, k8sServiceListenerIP, int(event.port)); err != nil {
-						log.Errorw("failed to create listener", log.Fields{
-							"error":     err,
-							"port":      event.port,
-							"namespace": event.namespace,
-							"name":      event.name,
-						})
+					if enablePrivilegedService {
+						portMapping, err := createPortMapping(event.portMapping, k8sServiceListenerIP)
+						if err != nil {
+							log.Errorw("failed to create port mapping", log.Fields{
+								"error":     err,
+								"ports":     event.portMapping,
+								"namespace": event.namespace,
+								"name":      event.name,
+							})
 
-						continue
+							continue
```

Comment:
I would really want to avoid `else if` 😱 . I will be refactoring this file here: https://github.com/rancher-sandbox/rancher-desktop-agent/issues/30. This should all go away soon.